pinot-transport/src/main/java/com/linkedin/pinot/routing/TableConfigRoutingTableSelector.java

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.routing;
+
+import com.google.common.base.Splitter;
+import com.linkedin.pinot.common.config.RealtimeTableConfig;
+import com.linkedin.pinot.common.metadata.ZKMetadataProvider;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import org.I0Itec.zkclient.IZkDataListener;
+import org.apache.commons.configuration.Configuration;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Routing table selector that uses the "routing.llc.percentage" (between 0-100) custom config in metadata to determine
+ * the appropriate ratio of queries to route to LLC.
+ */
+public class TableConfigRoutingTableSelector implements RoutingTableSelector, IZkDataListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TableConfigRoutingTableSelector.class);
+  private static final String LLC_ROUTING_PERCENTAGE_CONFIG_KEY = "routing.llc.percentage";
+  private static final long RATIO_MAP_ENTRY_MAX_STALENESS_MILLIS = TimeUnit.MINUTES.toMillis(10);
+  private static final Random RANDOM = new Random();
+
+  private final ConcurrentHashMap<String, RoutingRatio> _tableRoutingRatioMap = new ConcurrentHashMap<>();
+  private ZkHelixPropertyStore<ZNRecord> _propertyStore;
+
+  private boolean shouldUpdateRatioMapEntry(String realtimeTableName) {
+    // Always update ratio map entry if the table name is not there
+    if (!_tableRoutingRatioMap.containsKey(realtimeTableName)) {
+      return true;
+    }
+
+    // Update ratio map entry if it is stale
+    final RoutingRatio routingRatio = _tableRoutingRatioMap.get(realtimeTableName);
+    if (routingRatio._lastUpdateTime + RATIO_MAP_ENTRY_MAX_STALENESS_MILLIS < System.currentTimeMillis()) {
+      synchronized (routingRatio) {
+        return !routingRatio._isBeingUpdated;
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public boolean shouldUseLLCRouting(String realtimeTableName) {
+    if (shouldUpdateRatioMapEntry(realtimeTableName)) {
```

Comment:
The ZK read happens only if the value is stale or doesn't exist, otherwise it's just using the cached value. I could switch it to an async read if you think it's necessary.